### ISSUE TITLE
Question about the Loss Module

### ISSUE THREAD

**Issue by BIGKnight** [fla-org/flash-linear-attention#539]
Hi there,

Thank you for your excellent work on flash‑linear‑attention. While reviewing the implementation of FusedLinearCrossEntropy (see [here](https://github.com/fla-org/flash-linear-attention/blob/973b238275a2c477ec318d70cf657a1497382edf/fla/modules/fused_linear_cross_entropy.py#L370C9-L370C32)), I noticed that the reduction argument is set to "mean" by default. This choice behaves correctly when gradient accumulation is not used. However, when the accumulation step is greater than 1, averaging within a single forward pass may yield sub‑optimal gradients. 

Hugging Face discusses this subtlety in their gradient‑accumulation guide: [discussion](https://huggingface.co/blog/gradient_accumulation). In their implementation, the loss is first reduced with "sum" and only divided by the total token count after all accumulated steps, ensuring proper scaling. Would it make sense to adopt the same strategy here (i.e., switch the default reduction to "sum" and perform the final normalization after accumulation) or perhaps add a note in the documentation to highlight this behavior? 

The following is the key point quoted from [discussion](https://huggingface.co/blog/gradient_accumulation).
> To be precise, for gradient accumulation across token-level tasks like causal LM training, the correct loss should be computed by the total loss across all batches in a gradient accumulation step, divided by the total number of all non padding tokens in those batches. This is not the same as the average of the per-batch loss values.

Thank you again for the great library and for considering this suggestion.

**Comment by sustcsonglin** [comment]
Hi @BIGKnight, thanks for pointing this out. Do you think it might be better to return the sum of the cross-entropy loss and leave the normalization to the user in the training loop?

**Comment by BIGKnight** [comment]
Hi songlin

I believe this is a reasonable choice, particularly when gradient accumulation is enabled. 

My concern is that with reduction='mean', the loss is averaged within each micro‑batch, and these micro‑batch means are then averaged across accumulation steps. When micro‑batches contain unequal numbers of tokens, this two‑stage procedure is not equivalent to computing a single token‑level mean—i.e., summing losses over all tokens within the accumulation window and dividing by the total token count—so small discrepancies may arise. 

Here is my current modification, which is slightly based on the [modeling_delta_net.py：L390-L405](https://github.com/fla-org/flash-linear-attention/blob/4335842540c4445fcf12e913515bae2705b8ce47/fla/models/delta_net/modeling_delta_net.py#L389C9-L405C81):

`if labels is not None:

        num_items = kwargs.pop('num_items_in_batch', None)
        reduction = "sum" if num_items is not None and num_items > 1 else "mean"
        if getattr(self, 'criterion', None) is None:
            if fuse_linear_and_cross_entropy:
                criterion = FusedLinearCrossEntropyLoss(use_l2warp=self.config.use_l2warp, reduction=reduction)
            elif self.config.fuse_cross_entropy:
                criterion = FusedCrossEntropyLoss(inplace_backward=True, reduction=reduction)
            else:
                criterion = nn.CrossEntropyLoss()
        else:
            criterion = self.criterion
        labels = labels.to(hidden_states.device)
        labels = torch.cat((labels[..., 1:], torch.full_like(labels[:, :1], criterion.ignore_index)), 1)
        if fuse_linear_and_cross_entropy:
            loss = criterion(hidden_states, labels, self.lm_head.weight, self.lm_head.bias)
            loss = loss / num_items if num_items is not None and num_items > 1 else loss
        else:
            loss = criterion(logits.view(labels.numel(), -1), labels.view(-1))
            loss = l2_warp(loss, logits) if self.config.use_l2warp else loss
            loss = loss / num_items if num_items is not None and num_items > 1 else loss`

However, my experience in this area is limited, and I may have overlooked certain considerations, so it's be better for you authors to decide whether any changes are appropriate.

Lastly, thank you very much for open‑sourcing the project and for your helpful responses—FLA has been a pleasure to use and is highly efficient ： ）
